src/essence/Tools/Isochrone/IsochroneTool.js

```diff
@@ -0,0 +1,539 @@
+                    minResolution, //number
+                    maxResolution, //number
+                    zoomOffset //number
+                });
+            }
+        }
+    },
+
+    make: function() {
+        this.MMGISInterface = new interfaceWithMMGIS();
+        
+        $("#iscNew").on("click", () => this.addIsochrone());
+        this.containerEl = $("#isochroneOptionsContainer");
+        this.addIsochrone();
+    },
+
+    destroy: function() {
+        for(const {marker, layerName} of this.managers) {
+            Map_.rmNotNull(marker);
+            Map_.rmNotNull(L_.layersGroup[layerName]);
+        }
+
+        if(this.hoverPolyline !== null) {
+            this.hoverPolyline.remove(Map_.map);
+        }
+        
+        this.managers = [];
+        this.activeManager = null;
+        this.managerCounter = 0;
+        this.MMGISInterface.separateFromMMGIS();
+    },
+
+    getUrlString: function() { //TODO?
+        return '';
+    },
+
+    valueToColor: function(val, rampIndex, steps = 0) {
+        val = Math.min(val, 1);
+        if(steps) val = Math.floor(val * steps) / steps;
+
+        const ramp = this.colorRamps[rampIndex];
+        const color = val * (ramp.length - 1);
+        const i = Math.min(Math.floor(color), ramp.length - 2);
+        const off = color % 1;
+        const getChan = chan =>
+            Math.floor(ramp[i][chan] * (1 - off) + ramp[i + 1][chan] * off);
+        return [getChan(0), getChan(1), getChan(2)];
+    },
+
+    makeGradientEls: function() {
+        const C_WIDTH = 120, C_HEIGHT = 25;
+        const numRamps = this.colorRamps.length;
+        let colorEls = [];
+        for(let i = 0; i < numRamps; i++) {
+            let canvas = document.createElement("canvas");
+            canvas.width = C_WIDTH;
+            canvas.height = C_HEIGHT;
+            let ctx = canvas.getContext("2d");
+            let image = ctx.getImageData(0, 0, C_WIDTH, C_HEIGHT);
+            let data = image.data;
+            for(let x = 0; x < C_WIDTH; x++) {
+                const color = this.valueToColor(x / C_WIDTH, i);
+                for(let y = 0; y < C_HEIGHT; y++) {
+                    const di = (y * C_WIDTH * 4) + (x * 4);
+                    data[di] = color[0];
+                    data[di + 1] = color[1];
+                    data[di + 2] = color[2];
+                    data[di + 3] = 255;
+                }
+            }
+            ctx.putImageData(image, 0, 0);
+            colorEls.push(canvas);
+        }
+        return colorEls;
+    },
+
+    managerOnChange: function(manager) {
+        this.makeMarker(manager, true);
+        this.makeDataLayer(manager);
+    },
+    managerOnFocus: function(manager) {
+        if(this.activeManager === manager) return;
+        if(this.activeManager !== null) {
+            this.activeManager.unfocus();
+        }
+        this.activeManager = manager;
+        this.activeManager.focus();
+    },
+    managerOnDelete: function(manager) {
```

Comment:
remove empty functions. if the method must remain due to some class extension reason, then document the reason in a jsdoc style comment, eg.

```
/** Keep this empty function because blah blah blah */
managerOnDelete: function (manager) { }
```